Sources/NIOHPACK/HeaderTables.swift

```diff
@@ -0,0 +1,326 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Stores offsets into a contiguous buffer locating the name and value bytes
+/// for a header name and value pair.
+struct HeaderTableEntry {
+    var name: HPACKHeaderIndex
+    var value: HPACKHeaderIndex
+    
+    init(name: HPACKHeaderIndex, value: HPACKHeaderIndex) {
+        self.name = name
+        self.value = value
+    }
+    
+    // RFC 7541 § 4.1:
+    //
+    //      The size of an entry is the sum of its name's length in octets (as defined in
+    //      Section 5.2), its value's length in octets, and 32.
+    //
+    //      The size of an entry is calculated using the length of its name and value
+    //      without any Huffman encoding applied.
+    var length: Int {
+        return self.name.length + self.value.length + 32
+    }
+    
+    fileprivate mutating func adjust(by delta: Int, wrappingAt max: Int) {
+        name.adjust(by: delta, wrappingAt: max)
+        value.adjust(by: delta, wrappingAt: max)
+    }
+}
+
+extension HeaderTableEntry : CustomStringConvertible {
+    var description: String {
+        return "(name: \(self.name), value: \(self.value))"
+    }
+}
+
+/// Raw bytes storage for the header tables, both static and dynamic. Similar in spirit to
+/// `HPACKHeaders` and `NIOHTTP1.HTTPHeaders`, but uses a ring buffer to hold the bytes to
+/// avoid allocation churn while evicting and replacing entries.
+struct HeaderTableStorage {
+    static let defaultMaxSize = 4096
+    
+    private var buffer: StringRing
+    private var headers: CircularBuffer<HeaderTableEntry>
+    
+    private(set) var maxSize: Int
+    private(set) var length: Int = 0
+    
+    var count: Int {
+        return self.headers.count
+    }
+    
+    init(allocator: ByteBufferAllocator, maxSize: Int = HeaderTableStorage.defaultMaxSize) {
+        self.maxSize = maxSize
+        self.buffer = StringRing(allocator: allocator, capacity: self.maxSize)
+        self.headers = CircularBuffer(initialRingCapacity: self.maxSize / 64)    // rough guess: 64 bytes per header
+    }
+    
+    init(allocator: ByteBufferAllocator, staticHeaderList: [(String, String)]) {
+        // calculate likely total byte length we will actually need
+        // the static table is all ASCII, so character count == byte count here
+        let bytesNeeded = staticHeaderList.reduce(0) { $0 + $1.0.count + $1.1.count }
+        // set this according to what the length algorithm would normally expect
```

Comment:
This is a weird comment...what's it doing?

Reply:
Oops, dangling comment I think.